pkg/controller/oam/applicationconfiguration/render.go

```diff
@@ -96,26 +107,117 @@ func (r *components) Render(ctx context.Context, ac *v1alpha2.ApplicationConfigu
+			setTraitProperties(t, acc.ComponentName, ac.GetNamespace(), ref)
+
+			traits[i] = *t
+
+			traitDef, err := r.getTraitDefinition(ctx, t)
+			if err != nil {
+				return nil, err
 			}
+			traitDefs = append(traitDefs, traitDef)
+		}
+		if err := SetWorkloadInstanceName(traitDefs, w, c); err != nil {
+			return nil, err
+		}
+		workloads[i] = Workload{ComponentName: acc.ComponentName, ComponentRevisionName: componentRevisionName, Workload: w, Traits: traits}
+	}
+	return workloads, nil
+}
 
-			t.SetOwnerReferences([]metav1.OwnerReference{*ref})
-			t.SetNamespace(ac.GetNamespace())
+func setTraitProperties(t *unstructured.Unstructured, componentName, namespace string, ref *metav1.OwnerReference) {
+	// Set metadata name for `Trait` if the metadata name is NOT set.
+	if t.GetName() == "" {
+		t.SetName(componentName)
+	}
 
-			traits[i] = *t
+	t.SetOwnerReferences([]metav1.OwnerReference{*ref})
+	t.SetNamespace(namespace)
+}
+
+func (r *components) getTraitDefinition(ctx context.Context, t *unstructured.Unstructured) (v1alpha2.TraitDefinition, error) {
+	traitDefinitionName := getCRDName(t)
+	traitDef := v1alpha2.TraitDefinition{}
+	err := r.client.Get(ctx, types.NamespacedName{
+		Name: traitDefinitionName,
+	}, &traitDef)
+	if err != nil {
+		return traitDef, errors.Wrapf(err, errFmtGetTraitDefinition, traitDefinitionName)
+	}
+	return traitDef, nil
+}
+
+// SetWorkloadInstanceName will set metadata.name for workload CR according to createRevision flag in traitDefinition
+func SetWorkloadInstanceName(traitDefs []v1alpha2.TraitDefinition, w *unstructured.Unstructured, c *v1alpha2.Component) error {
+	//Don't override the specified name
+	if w.GetName() != "" {
+		return nil
+	}
+	pv := fieldpath.Pave(w.Object)
+	if IsRevisionEnabled(traitDefs) {
+		// if revisionEnabled, use revisionName as workload name
+		if err := pv.SetString(instanceNamePath, c.Status.LatestRevision.Name); err != nil {
+			return errors.Wrapf(err, errSetValueForField, instanceNamePath, c.Status.LatestRevision)
+		}
+		return nil
+	}
+	// use component name as workload name, which means we will always use one workload for different revisions
+	if err := pv.SetString(instanceNamePath, c.GetName()); err != nil {
+		return errors.Wrapf(err, errSetValueForField, instanceNamePath, c.GetName())
+	}
+	w.Object = pv.UnstructuredContent()
+	return nil
+}
+
+// IsRevisionEnabled will check if any trait has createRevision flag, the appconfig should create a new workload instance
+func IsRevisionEnabled(traitDefs []v1alpha2.TraitDefinition) bool {
+	for _, td := range traitDefs {
+		if td.Spec.RevisionEnabled {
+			return true
 		}
+	}
+	return false
+}
 
-		workloads[i] = Workload{ComponentName: acc.ComponentName, Workload: w, Traits: traits}
+func getCRDName(u *unstructured.Unstructured) string {
+	group, _ := addon.ApiVersion2GroupVersion(u.GetAPIVersion())
+	resources := []string{addon.Kind2Resource(u.GetKind())}
+	if group != "" {
+		resources = append(resources, group)
 	}
```

Comment:
I think there is a new function called GetGVResource that one can use now instead of calling two functions.